spring-cloud-task-batch/src/main/java/org/springframework/cloud/task/batch/configuration/TaskBatchExecutionListenerFactoryBean.java

```diff
@@ -0,0 +1,119 @@
+ * limitations under the License.
+ */
+package org.springframework.cloud.task.batch.configuration;
+
+import java.lang.reflect.Field;
+import javax.sql.DataSource;
+
+import org.springframework.aop.framework.Advised;
+import org.springframework.aop.support.AopUtils;
+import org.springframework.beans.factory.FactoryBean;
+import org.springframework.cloud.task.batch.listener.TaskBatchExecutionListener;
+import org.springframework.cloud.task.batch.listener.support.JdbcTaskBatchDao;
+import org.springframework.cloud.task.batch.listener.support.MapTaskBatchDao;
+import org.springframework.cloud.task.repository.TaskExplorer;
+import org.springframework.cloud.task.repository.dao.MapTaskExecutionDao;
+import org.springframework.cloud.task.repository.support.SimpleTaskExplorer;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.util.Assert;
+import org.springframework.util.ReflectionUtils;
+import org.springframework.util.StringUtils;
+
+/**
+ * {@link FactoryBean} for a {@link TaskBatchExecutionListener}.  Provides a jdbc based
+ * listener if there is a {@link DataSource} available.  Otherwise, builds a listener that
+ * uses the map based implementation.
+ *
+ * @author Michael Minella
+ */
+public class TaskBatchExecutionListenerFactoryBean implements FactoryBean<TaskBatchExecutionListener> {
+
+	private ConfigurableApplicationContext context;
+
+	private TaskBatchExecutionListener listener;
+
+	private String dataSourceName;
+
+	/**
+	 * @param context the current application context
+	 */
+	public TaskBatchExecutionListenerFactoryBean(ConfigurableApplicationContext context) {
+		Assert.notNull(context, "A ConfigurableApplicationContext is required");
+
+		this.context = context;
+	}
+
+	@Override
+	public TaskBatchExecutionListener getObject() throws Exception {
+		if(this.context.getBeanNamesForType(DataSource.class).length == 0) {
+			this.listener = new TaskBatchExecutionListener(getMapTaskBatchDao());
+		}
+		else {
+			DataSource dataSource;
+
+			if(StringUtils.hasText(this.dataSourceName)) {
+				dataSource = (DataSource) this.context.getBean(this.dataSourceName);
+			}
+			else {
+				if(this.context.getBeanNamesForType(DataSource.class).length == 1) {
+					dataSource = this.context.getBean(DataSource.class);
+				}
+				else {
+					throw new IllegalStateException("Unable to determine what DataSource to use");
+				}
+			}
+			this.listener = new TaskBatchExecutionListener(new JdbcTaskBatchDao(dataSource));
+		}
+
+		return listener;
+	}
+
+	@Override
+	public Class<?> getObjectType() {
+		return TaskBatchExecutionListener.class;
+	}
+
+	@Override
+	public boolean isSingleton() {
+		return true;
+	}
+
+	public void setDataSourceName(String dataSourceName) {
+		this.dataSourceName = dataSourceName;
+	}
+
+	private MapTaskBatchDao getMapTaskBatchDao() throws Exception {
+		Field taskExecutionDaoField = ReflectionUtils.findField(SimpleTaskExplorer.class, "taskExecutionDao");
```

Comment:
Nevermind.  Its MapTaskExecutionDao, they should only be using simple.  (Answered my own question)